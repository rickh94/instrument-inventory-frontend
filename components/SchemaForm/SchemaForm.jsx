import PropTypes from 'prop-types'
import React from 'react'

import Ajv from 'ajv'
import { JSONSchemaBridge } from 'uniforms-bridge-json-schema'
import {
  AutoField,
  AutoFields,
  AutoForm,
  ErrorsField,
  SubmitField,
  TextField,
} from 'uniforms-material'
import { Button, FormGroup } from '@material-ui/core'
import { makeStyles } from '@material-ui/styles'
import clsx from 'clsx'

import SchemaRatingField from '../SchemaRatingField'
import SchemaScannerField from '../SchemaScannerField'
import { lastButton } from '../../globalStyles'

const ajv = new Ajv({ allErrors: true, useDefaults: true })

const createValidator = schema => {
  const validator = ajv.compile(schema)

  return model => {
    validator(model)
    if (validator.errors && validator.errors.length) {
      throw { details: validator.errors }
    }
  }
}

const useStyles = makeStyles(theme => ({
  button: {
    margin: theme.spacing(1),
  },
  lastButton,
}))

const SchemaForm = ({
  schema,
  initialData,
  omitFields = [],
  onSubmit,
  error,
  onChange,
  onCancel,
}) => {
  const schemaValidator = createValidator(schema)
  const bridge = new JSONSchemaBridge(schema, schemaValidator)
  if (schema.properties.password) {
    omitFields.push('password')
  }
  if (schema.properties.confirm_password) {
    omitFields.push('confirm_password')
  }
  if (schema.properties.number) {
    schema.properties.number.uniforms = {
      component: SchemaScannerField,
    }
  }
  if (schema.properties.quality) {
    schema.properties.quality.uniforms = {
      component: SchemaRatingField,
    }
  }
  if (schema.properties.condition) {
    schema.properties.condition.uniforms = {
      component: SchemaRatingField,
    }
  }

  if (initialData) {
    Object.keys(schema.properties).forEach(key => {
      if (initialData[key] === null || initialData[key] === undefined) {
        if (schema.properties[key].type === 'string') {
          initialData[key] = ''
        } else if (schema.properties[key].type === 'integer') {
          initialData[key] = 0
        }
      }
    })
  }

  const classes = useStyles()
  return (
    <AutoForm
      schema={bridge}
      onSubmit={onSubmit}
      error={error}
      showInlineError
      onChange={onChange}
      model={initialData}
      data-testid="schema-form"
    >
      <AutoFields omitFields={omitFields} />
      {schema.properties.password && (
        <AutoField component={TextField} type="password" name="password" />
      )}
      {schema.properties.confirm_password && (
        <AutoField component={TextField} type="password" name="confirm_password" />
      )}
      <ErrorsField />
      <FormGroup row>
        {onCancel && (
          <Button
            onClick={onCancel}
            variant="text"
            className={clsx(classes.button, classes.lastButton)}
          >
            Cancel
          </Button>
        )}
        <SubmitField
          color="primary"
          variant="text"
          className={clsx(classes.button, !onCancel && classes.lastButton)}
        />
      </FormGroup>
    </AutoForm>
  )
}

SchemaForm.propTypes = {
  schema: PropTypes.object,
  initialData: PropTypes.object,
  omitFields: PropTypes.arrayOf(PropTypes.string),
  onSubmit: PropTypes.func.isRequired,
  onChange: PropTypes.func,
  onCancel: PropTypes.func,
  error: PropTypes.object,
}

export default SchemaForm
