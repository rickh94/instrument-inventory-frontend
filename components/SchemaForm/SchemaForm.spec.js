import React from 'react'
import SchemaForm from './SchemaForm'
import { render, cleanup, fireEvent } from '@testing-library/react'
import { MuiThemeProvider, createMuiTheme } from '@material-ui/core'

afterEach(cleanup)

describe('<SchemaForm />', () => {
  test('renders', () => {
    const { container } = render(
      <MuiThemeProvider theme={createMuiTheme()}>
        <SchemaForm
          schema={{ properties: {} }}
          initialData={{}}
          onSubmit={jest.fn()}
          onChange={jest.fn()}
          onCancel={jest.fn()}
        />
      </MuiThemeProvider>
    )
    expect(container).toMatchSnapshot()
  })

  test('renders schema', () => {
    const testSchema = {
      type: 'object',
      properties: {
        testNumber: {
          title: 'Test Number',
          description: 'Number',
          type: 'number',
        },
        stuff: {
          title: 'Stuff',
          description: 'A Text Input',
          type: 'string',
        },
        choice: {
          title: 'Choice',
          description: 'Some choices',
          enum: ['a1', 'b2', 'c3', 'd4'],
        },
      },
    }
    const { queryByText, container } = render(
      <MuiThemeProvider theme={createMuiTheme()}>
        <SchemaForm
          schema={testSchema}
          onSubmit={jest.fn()}
          onChange={jest.fn()}
          onCancel={jest.fn()}
        />
      </MuiThemeProvider>
    )
    expect(queryByText('Test Number')).toBeTruthy()
    expect(queryByText('Stuff')).toBeTruthy()
    expect(queryByText('Choice')).toBeTruthy()

    expect(container.querySelector('input[name="stuff"]')).toBeTruthy()
    expect(container.querySelector('input[name="stuff"]').type).toEqual('text')

    expect(container.querySelector('input[name="testNumber"]')).toBeTruthy()
    expect(container.querySelector('input[name="testNumber"]').type).toEqual('number')

    expect(container.querySelector('input[name="choice"]')).toBeTruthy()
    expect(container.querySelector('input[name="choice"]').type).toEqual('hidden')

    expect(queryByText('Submit')).toBeTruthy()
    expect(queryByText('Cancel')).toBeTruthy()
  })

  test('renders a scanner field for a number in schema', () => {
    const testSchema = {
      type: 'object',
      properties: {
        number: {
          title: 'Number',
          description: 'Scanning Number',
          type: 'string',
        },
      },
    }
    const { queryByText, container } = render(
      <MuiThemeProvider theme={createMuiTheme()}>
        <SchemaForm schema={testSchema} onSubmit={jest.fn()} onChange={jest.fn()} />
      </MuiThemeProvider>
    )
    expect(queryByText('Number')).toBeTruthy()

    expect(queryByText('Scan Barcode')).toBeTruthy()
    expect(container.querySelector('input#scanner-field')).toBeTruthy()

    expect(queryByText('Submit')).toBeTruthy()
  })

  test('renders cancel if onCancel prop, and calls', () => {
    const testSchema = {
      type: 'object',
      properties: {
        stuff: {
          title: 'Stuff',
          description: 'A Text Input',
          type: 'string',
        },
      },
    }
    const onCancel = jest.fn()
    const { queryByText } = render(
      <MuiThemeProvider theme={createMuiTheme()}>
        <SchemaForm
          schema={testSchema}
          onSubmit={jest.fn()}
          onChange={jest.fn()}
          onCancel={onCancel}
        />
      </MuiThemeProvider>
    )
    expect(queryByText('Cancel')).toBeTruthy()
    fireEvent.click(queryByText('Cancel'))
    expect(onCancel).toHaveBeenCalled()
  })

  test('does not render cancel if no onCancel prop', () => {
    const testSchema = {
      type: 'object',
      properties: {
        stuff: {
          title: 'Stuff',
          description: 'A Text Input',
          type: 'string',
        },
      },
    }
    const { queryByText } = render(
      <MuiThemeProvider theme={createMuiTheme()}>
        <SchemaForm schema={testSchema} onSubmit={jest.fn()} onChange={jest.fn()} />
      </MuiThemeProvider>
    )
    expect(queryByText('Cancel')).toBeFalsy()
  })

  test('calls onSubmit with correct values', async () => {
    const testSchema = {
      type: 'object',
      properties: {
        testNumber: {
          title: 'Test Number',
          description: 'Number',
          type: 'number',
        },
        stuff: {
          title: 'Stuff',
          description: 'A Text Input',
          type: 'string',
        },
        choice: {
          title: 'Choice',
          description: 'Some choices',
          enum: ['a1', 'b2', 'c3', 'd4'],
        },
      },
    }
    const onSubmit = jest.fn()
    const { container } = render(
      <MuiThemeProvider theme={createMuiTheme()}>
        <SchemaForm
          schema={testSchema}
          onSubmit={onSubmit}
          onChange={jest.fn()}
          onCancel={jest.fn()}
        />
      </MuiThemeProvider>
    )
    fireEvent.change(container.querySelector('input[name="stuff"]'), {
      target: { value: 'stuff value' },
    })
    fireEvent.change(container.querySelector('input[name="testNumber"]'), {
      target: { value: 5 },
    })

    fireEvent.click(container.querySelector('button[type="submit"]'))

    // eslint-disable-next-line no-undef
    await new Promise(resolve => process.nextTick(resolve))

    expect(onSubmit).toHaveBeenCalledWith({
      stuff: 'stuff value',
      testNumber: 5,
    })
  })

  test('renders password and confirm password as password fields', () => {
    const testSchema = {
      type: 'object',
      properties: {
        password: {
          title: 'Password',
          description: 'User Password',
          type: 'string',
        },
        confirm_password: {
          title: 'Confirm Password',
          description: 'Confirm User Password',
          type: 'string',
        },
      },
    }
    const { queryByText, container } = render(
      <MuiThemeProvider theme={createMuiTheme()}>
        <SchemaForm
          schema={testSchema}
          onSubmit={jest.fn()}
          onChange={jest.fn()}
          onCancel={jest.fn()}
        />
      </MuiThemeProvider>
    )
    expect(queryByText('Password')).toBeTruthy()
    expect(queryByText('Confirm Password')).toBeTruthy()

    expect(container.querySelector('input[name="password"]')).toBeTruthy()
    expect(container.querySelector('input[name="password"]').type).toEqual('password')

    expect(container.querySelector('input[name="confirm_password"]')).toBeTruthy()
    expect(container.querySelector('input[name="confirm_password"]').type).toEqual(
      'password'
    )
  })

  test('renders rating field for quality and condition', () => {
    const testSchema = {
      type: 'object',
      properties: {
        quality: {
          title: 'Quality',
          type: 'number',
        },
        condition: {
          title: 'Condition',
          type: 'number',
        },
      },
    }
    const { queryByText, container } = render(
      <MuiThemeProvider theme={createMuiTheme()}>
        <SchemaForm
          schema={testSchema}
          onSubmit={jest.fn()}
          onChange={jest.fn()}
          onCancel={jest.fn()}
        />
      </MuiThemeProvider>
    )

    expect(queryByText('Quality')).toBeTruthy()
    expect(queryByText('Condition')).toBeTruthy()
    expect(container.querySelectorAll('.Rating-root-646').length).toEqual(2)
  })
})
