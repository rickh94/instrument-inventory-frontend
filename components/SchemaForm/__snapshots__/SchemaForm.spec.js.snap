// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SchemaForm /> renders 1`] = `
<div>
  <form
    data-testid="schema-form"
    novalidate=""
  >
    <div />
    <div
      class="MuiFormGroup-root MuiFormGroup-row"
    >
      <button
        class="MuiButtonBase-root MuiButton-root makeStyles-button-1 makeStyles-lastButton-2 MuiButton-text"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Cancel
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root makeStyles-button-1 MuiButton-textPrimary MuiButton-text"
        tabindex="0"
        type="submit"
      >
        <span
          class="MuiButton-label"
        >
          Submit
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </form>
</div>
`;
